lxd/containers.go

```diff
@@ -51,90 +54,227 @@ func containersPost(d *Daemon, r *http.Request) Response {
 	if err != nil {
+		removeContainerPath(name)
 		return InternalError(err)
 	}
 
 	/*
-	 * Set the id mapping. This may not be how we want to do it, but it's a
-	 * start.  First, we remove any id_map lines in the config which might
-	 * have come from ~/.config/lxc/default.conf.  Then add id mapping based
-	 * on Domain.id_map
+	 * extract the rootfs asynchronously
 	 */
-	if d.id_map != nil {
-		shared.Debugf("setting custom idmap")
-		err = c.SetConfigItem("lxc.id_map", "")
-		if err != nil {
-			shared.Debugf("Failed to clear id mapping, continuing")
-		}
-		uidstr := fmt.Sprintf("u 0 %d %d\n", d.id_map.Uidmin, d.id_map.Uidrange)
-		shared.Debugf("uidstr is %s\n", uidstr)
-		err = c.SetConfigItem("lxc.id_map", uidstr)
-		if err != nil {
-			return InternalError(err)
-		}
-		gidstr := fmt.Sprintf("g 0 %d %d\n", d.id_map.Gidmin, d.id_map.Gidrange)
-		err = c.SetConfigItem("lxc.id_map", gidstr)
-		if err != nil {
-			return InternalError(err)
-		}
+	run := shared.OperationWrap(func() error { return extractShiftRootfs(uuid, name, d) })
+	return &asyncResponse{run: run, containers: []string{req.Name}}
+}
+
+func removeContainerPath(name string) {
+	cpath := shared.VarPath("lxc", name)
+	err := os.RemoveAll(cpath)
+	if err != nil {
+		shared.Debugf("Error cleaning up %s: %s\n", cpath, err)
+	}
+}
+
+func extractShiftRootfs(uuid string, name string, d *Daemon) error {
+	cleanup := func(err error) error {
+		removeContainerPath(name)
+		dbRemoveContainer(name)
+		return err
 	}
 
 	/*
-	 * Actually create the container
+	 * We want to use archive/tar for this, but that doesn't appear
+	 * to be working for us (see lxd/images.go)
+	 * So for now, we extract the rootfs.tar.xz from the image
+	 * tarball to /var/lib/lxd/lxc/container/rootfs.tar.xz, then
+	 * extract that under /var/lib/lxd/lxc/container/rootfs/
 	 */
-	run := shared.OperationWrap(func() error { return c.Create(opts) })
-	return &asyncResponse{run: run, containers: []string{req.Name}}
+	dpath := shared.VarPath("lxc", name)
+	fmt.Printf("uuid is %s\n", uuid)
+	imagefile := shared.VarPath("images", uuid)
+	output, err := exec.Command("tar", "-C", dpath, "-Jxf", imagefile, "rootfs.tar.xz").Output()
+	if err != nil {
+		fmt.Printf("Untar of image: Output %s\nError %s\n", output, err)
+		return cleanup(err)
+	}
+
+	rpath := shared.VarPath("lxc", name, "rootfs")
+	tarfile := shared.VarPath("lxc", name, "rootfs.tar.xz")
+	output, err = exec.Command("tar", "-C", rpath, "--numeric-owner", "-Jxpf", tarfile).Output()
+	if err != nil {
+		fmt.Printf("Untar of rootfs: Output %s\nError %s\n", output, err)
+		return cleanup(err)
+	}
+
+	err = d.id_map.ShiftRootfs(rpath)
+	if err != nil {
+		fmt.Printf("Shift of rootfs %s failed: %s\n", rpath, err)
+		return cleanup(err)
+	}
+
+	/* Set an acl so the container root can descend the container dir */
+	acl := fmt.Sprintf("%d:rx", d.id_map.Uidmin)
+	_, err = exec.Command("setfacl", "-m", acl, dpath).Output()
+	if err != nil {
```

Comment:
We should document that setfacl and tar are now dependencies of LXD (probably in README).